### PR TITLE
check docId vs lastDocId

### PR DESCRIPTION
**Pull request by ashtul**
fix crash on #1282 but not wrong results

### CODE REVIEW

File: src/pytest/test_issues.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  env.expect('FT.ADD idx doc1 1.0 FIELDS txt1 foo txt2 bar').equal('OK')
+  env.expect('FT.ADD idx doc2 1.0 FIELDS txt1 baz txt2 xyz').equal('OK')
+  env.expect('FT.ADD idx doc3 1.0 FIELDS txt1 hello txt2 world').equal('OK')
+  env.expect('FT.ADD idx doc4 1.0 FIELDS txt1 jim txt2 burie').equal('OK')
+
+
+  # test with tags would crash server
+  env.expect('FT.SEARCH idx', '~world ~bar')
+
+#  res = [4L, 'doc3', ['txt1', 'hello', 'txt2', 'world'],
```
Comment (MeirShpilraien) [inline src/pytest/test_issues.py:12]: Remove?

File: src/index.c [hunk 2]
```diff
@@ -978,6 +977,7 @@ static IndexCriteriaTester *NI_GetCriteriaTester(void *ctx) {
 static size_t NI_NumEstimated(void *ctx) {
   NotContext *nc = ctx;
   return nc->maxDocId;
+  return SIZE_MAX;
```
Comment (MeirShpilraien) [inline src/index.c:980]: Why is this needed?

File: src/id_list.c [hunk 3]
```diff
@@ -187,6 +187,5 @@ IndexIterator *NewIdListIterator(t_docId *ids, t_offset num, double weight) {
   ret->mode = MODE_SORTED;
 
   ret->HasNext = NULL;
-  ret->GetCurrent = NULL;
```
Comment (MeirShpilraien) [inline src/id_list.c:190]: Why is it bad to initialize it?

File: src/index.c [hunk 4]
```diff
@@ -604,7 +604,6 @@ IndexIterator *NewIntersecIterator(IndexIterator **its_, size_t num, DocTable *d
   it->Free = IntersectIterator_Free;
   it->Abort = II_Abort;
   it->Rewind = II_Rewind;
-  it->GetCurrent = NULL;
```
Comment (MeirShpilraien) [inline src/index.c:607]: Same?

File: src/index.c (hunk 2 above)
Comment (ashtul) [reply src/index.c:980]: oops

File: src/pytest/test_issues.py (hunk 1 above)
Comment (ashtul) [reply src/pytest/test_issues.py:12]: We should fix it next.
Thought of it as TODO...

File: src/index_iterator.h [hunk 5]
```diff
@@ -34,8 +34,6 @@ typedef struct indexIterator {
 
   int mode;
 
-  RSIndexResult *(*GetCurrent)(void *ctx);
```
Comment (ashtul) [inline src/index_iterator.h:37]: @MeirShpilraien I have removed the pointer altogether as it isn't being used...

File: src/id_list.c (hunk 3 above)
Comment (ashtul) [reply src/id_list.c:190]: removed pointer

File: src/index.c (hunk 4 above)
Comment (ashtul) [reply src/index.c:607]: same

**Comment by ashtul** [comment]
> @ashtul can you explain why it's only crashing with tags, why you fix fixes it?

I cannot explain why is it happening only with tags.
The crash caused as the `docId` was advanced beyond `maxDocId` and `OI_SkipTo` tested against `nc->lastDocId`.
Next line is `nc->lastDocId = docId;` so it makes complete sense to check `docID` instead.

**Comment by MeirShpilraien** [comment]
We need to understand why it only happened with tags, maybe tags can not handle ids which is more than max docid and we will get this crash from another code path...

**Comment by MeirShpilraien** [comment]
@ashtul what about this one?

**Comment by ashtul** [comment]
@MeirShpilraien 
Crush is unrelated to tags but occurs when there is no index for the optional term.

**Comment by MeirShpilraien** [comment]
@ashtul not follow how come `if (docId > nc->maxDocId) {` fixes it?

**Comment by ashtul** [comment]
@MeirShpilraien 
Change the fix slightly. We haven't updated the context `lastId` immediately and used the older ID and didn't return `INDEXREAD_EOF` but an `RSIndexResult` with `NULL` which caused the crash.

**Comment by MeirShpilraien** [comment]
@ashtul sorry, don't see where you return INDEXREAD_EOF instead of NULL in the diff, can you point me to where is it?

**Comment by ashtul** [comment]
@gkorland 
fix crash on #1305

File: src/index.c [hunk 6]
```diff
@@ -1135,12 +1133,13 @@ static int OI_SkipTo(void *ctx, t_docId docId, RSIndexResult **hit) {
   //  nc->maxDocId, nc->lastDocId);
 
   int found = 0;
-  if (nc->lastDocId > nc->maxDocId) {
-    return INDEXREAD_EOF;
-  }
 
   // Set the current ID
   nc->lastDocId = docId;
```
Comment (ashtul) [inline src/index.c:1138]: @MeirShpilraien 
By setting `nc->lastDocId = docId` before `if`, we return `INDEXREAD_EOF` instead of an empty `hit`
